### ISSUE TITLE
n900 New kernel no longer finding internal mtd device, fremantle mounting etc.

### ISSUE THREAD

**Issue by SimonIremonger** [maemo-leste/bugtracker#265]
Last I checked, the Kernel and udev on leste on N900, can no longer mount the internal MTD unlike earlier builds.

See:-
https://leste.maemo.org/Nokia_N900#Accessing_MTD

I believe, the /dev/mtd5 etc. simply do not appear, and as such you cannot mount a fremantle or otherwise on the internal mtd!.  This may be as simple as a kernel config matter, or some different change in devicetree or otherwise.  Supposedly it was working on earlier leste N900 images/kernels...

**Comment by SimonIremonger** [comment]
@MerlijnWajer please take ownership as you had promised!

**Comment by MerlijnWajer** [comment]
This patch will probably fix it https://patchwork.kernel.org/patch/11317567/

**Comment by spinal84** [comment]
Please show the output of these two commands:
`uname -a
cat /proc/mtd`

**Comment by MerlijnWajer** [comment]
@spinal84 - above patch should really fix the problem, it's a regression in Linux AFAICT

**Comment by spinal84** [comment]
@MerlijnWajer, I couldn't reproduce the issue with both N900 kernels we have - in stable and devel repos, on fresh Maemo Leste install. The "fix" you mentioned is done in 2020, both our kernels are from 2019.

**Comment by MerlijnWajer** [comment]
In that case, here's my device report:

```
root@n900devuan:~# cat /proc/mtd
dev:    size   erasesize  name
root@n900devuan:~# uname -a
Linux n900devuan 5.1.0 #1 Fri Nov 1 17:33:03 UTC 2019 armv7l GNU/Linux
```

**Comment by MerlijnWajer** [comment]
And yeah, you're right, the timing is wrong on that one. It must be something else.

**Comment by spinal84** [comment]
Could you please paste the output of these commands after reboot:
```
dmesg | egrep -i 'mtd|command'
```
Or a better option: attaching full dmesg log to pastebin.

**Comment by SimonIremonger** [comment]
Basing upon the original leste image,   maemo-leste-1.0-armhf-n900-20190225.img.xz   ...

Quite clear the kernel is the problem!.  Search for "onenand" in the 3 dmesg'es, only working in the oldest (4.15.7) kernel!.
https://www.iremonger.me.uk/noidx/leste-20190225-onenand-test-kernels/

I note, the most recent kernel has a ONENAND VERIFY feature enabled in the kconfig, but I don't think thats' anything todo with the issue.

I have also tested the latest leste image 20200112 and the failure seems to be exactly the same, and can get /proc/mtd working by manually copying in the old 4.15.7 kernel modules and kernel-image, with extra uboot entry .....

**Comment by spinal84** [comment]
```
root@devuan-n900:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00020000 00020000 "bootloader"
mtd1: 00060000 00020000 "config"
mtd2: 00040000 00020000 "log"
mtd3: 00200000 00020000 "kernel"
mtd4: 00200000 00020000 "initfs"
mtd5: 0fb40000 00020000 "rootfs"
root@devuan-n900:~# uname -a
Linux devuan-n900 5.1.0 #1 Fri Nov 1 17:33:03 UTC 2019 armv7l GNU/Linux
```
Could you guys please send the full dmesg log after reboot to pastebin?

**Comment by SimonIremonger** [comment]
@spinal84   my dmesg'es and uname are linked above!
I NOTE mine is:-
Linux devuan-n900 5.1.0 #1 Fri Jul 26 07:51:50 UTC 2019 armv7l GNU/Linux
You must have an updated, fixed, kernel spinal84 ??   I wonder if any difference is, in part, the devicetree being supplied?

**Comment by SimonIremonger** [comment]
@spinal84 I have updated the above linked page -- trying 'k3' kernel from leste-devel too on latest image, similar failure.  @MerlijnWajer was going to pastebin 'dmesg' on n900, as they also get failure on Nov 1 kernel where you apparently do not!  I wonder if only certain N900s are affected or otherwise!.

**Comment by MerlijnWajer** [comment]
dmesg https://pastebin.com/raw/iwLjcNVw

**Comment by MerlijnWajer** [comment]
(misclick)

**Comment by SimonIremonger** [comment]
@spinal84 Please post your dmesg (where Nov 1 kernel mysteriously DOES get mtd devices) and your analysis of what you can see may be different in messages produced [do you have a different n900 hardware revision or anything?]...  My "k3" and MerlinWajer's boots have been fine ....    Attempt to narrow it down to specific combination of circumstances that triggers the failure!

**Comment by spinal84** [comment]
I got 2 more devices which behave like yours and the one of @MerlijnWajer.
I can debug it now. Seems like a kernel regression.

If you're curious, the first line in the working device's dmesg that differs is:
[    0.480224] OneNAND Manufacturer: Numonyx (0x20)
Compare to yours:
[    0.477539] OneNAND Manufacturer: Unknown (0xa0)

**Comment by MerlijnWajer** [comment]
It maybe worth checking if it simply works on a newer kernel.

**Comment by spinal84** [comment]
The issue should be fixed with linux-image-n900-5.1.21.1 which is uploaded to leste-devel repository.
Please upgrade and check. This will go to stable repo in a week or two.

**Comment by SimonIremonger** [comment]
Can confirm this is indeed fixed! =).  Closing bug.
